vm/src/stdlib/itertools.rs

```diff
@@ -908,48 +908,179 @@ impl PyItertoolsCombinations {
-        // If no suitable index is found, then the indices are all at
-        // their maximum value and we're done.
-        if !sentinel {
+        .into_ref_with_type(vm, cls)
+    }
+
+    #[pymethod(name = "__iter__")]
+    fn iter(zelf: PyRef<Self>, _vm: &VirtualMachine) -> PyRef<Self> {
+        zelf
+    }
+
+    #[pymethod(name = "__next__")]
+    fn next(&self, vm: &VirtualMachine) -> PyResult {
+        // stop signal
+        if self.exhausted.get() {
+            return Err(new_stop_iteration(vm));
+        }
+
+        let n = self.pool.len();
+        let r = self.r.get();
+
+        if n == 0 {
             self.exhausted.set(true);
+            return Ok(vm.ctx.new_tuple(vec![]));
         }
 
-        // Increment the current index which we know is not at its
-        // maximum.  Then move back to the right setting each index
-        // to its lowest possible value (one higher than the index
-        // to its left -- this maintains the sort order invariant).
-        indices[idx] += 1;
-        for j in idx + 1..r {
-            indices[j] = indices[j - 1] + 1;
+        let mut result = self.result.borrow_mut();
+
+        if result.is_none() {
+            // On the first pass, initialize result tuple using the indices
+            result.replace((0..r).collect());
+        } else {
+            let mut indices = self.indices.borrow_mut();
+            let mut cycles = self.cycles.borrow_mut();
+            let mut sentinel = false;
+
+            // Decrement rightmost cycle, moving leftward upon zero rollover
+            for i in (0..r).rev() {
+                cycles[i] -= 1;
+
+                if cycles[i] == 0 {
+                    // rotation: indices[i:] = indices[i+1:] + indices[i:i+1]
+                    let index = indices[i];
+                    for j in i..n - 1 {
+                        indices[j] = indices[j + i];
+                    }
+                    indices[n - 1] = index;
+                    cycles[i] = n - i;
+                } else {
+                    let j = cycles[i];
+                    let index = indices[i];
+                    indices[i] = indices[n - j];
+                    indices[n - j] = index;
+
+                    for k in i..r {
+                        // start with i, the leftmost element that changed
+                        // yield tuple(pool[k] for k in indices[:r])
+                        result.as_mut().unwrap()[k] = indices[k];
+                    }
+                    sentinel = true;
+                    break;
+                }
+            }
+            if !sentinel {
+                self.exhausted.set(true);
+                return Err(new_stop_iteration(vm));
+            }
         }
 
-        Ok(res.into_ref(vm).into_object())
+        Ok(vm.ctx.new_tuple(
+            result
+                .as_ref()
+                .unwrap()
+                .iter()
+                .map(|&i| self.pool[i].clone())
+                .collect(),
+        ))
```

Comment:
CPython tries to be more allocation-efficient by re-using the last-returned tuple (modifying it in-place) if the refcount == 1.  I'm not sure that it would be so straightforwards for us without making custom containers, unless I'm missing something?  Just wanted to point it out.

Reply:
I think you can use [`Rc::get_mut`](https://doc.rust-lang.org/std/rc/struct.Rc.html#method.get_mut) for this functionality.

Reply:
Or [`Rc::make_mut`](https://doc.rust-lang.org/std/rc/struct.Rc.html#method.make_mut) might be useful here as well?

Reply:
Oh, or you might want to add those sorta methods to `PyRef`, as you'd have to use the `Rc` functions on a `PyObjectRef`.

Reply:
Oh, or you might want to add those sorta methods to `PyRef`, as you'd have to use the `Rc` functions on a `PyObjectRef`.